File: 0438 Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        res = []
        if not s or not p or len(s) < len(p): return res
        pHash = {}
        for c in 'abcdefghijklmnopqrstuvwxyz':
            pHash[c] = 0
        for c in p:
            pHash[c] += 1
        cnt, l, r = len(p), 0, 0
        while r < len(s):
            if pHash[s[r]] > 0: cnt -= 1
            pHash[s[r]] -= 1
            r += 1
            if cnt == 0: res.append(l)
            if r - l == len(p):
                if pHash[s[l]] >= 0: cnt += 1
                pHash[s[l]] += 1
                l += 1
        return res

File: 0438 Find All Anagrams in a String.java
/* Adapted from NathanNi */
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> list = new ArrayList<>();
        if (s == null || s.length() == 0 || p == null || p.length() == 0 || s.length() < p.length()) return list;
        int[] hash = new int[26];
        for (char c : p.toCharArray()) {
            hash[c - 'a']++;
        }
        int left = 0, right = 0, count = p.length();
        while (right < s.length()) {
            if (hash[s.charAt(right++) - 'a']-- > 0) count--;
            if (count == 0) list.add(left);
            if (right - left == p.length() && hash[s.charAt(left++) - 'a']++ >= 0) count++;
        }
        return list;
    }
}